pkg/watermark/fetch/edge_fetcher.go

```diff
@@ -44,24 +45,37 @@ type edgeFetcher struct {
 	bucketName       string
 	storeWatcher     store.WatermarkStoreWatcher
 	processorManager *processor.ProcessorManager
+	lastProcessedWm  []int64
 	log              *zap.SugaredLogger
+	sync.Mutex
 }
 
 // NewEdgeFetcher returns a new edge fetcher.
-func NewEdgeFetcher(ctx context.Context, bucketName string, storeWatcher store.WatermarkStoreWatcher, manager *processor.ProcessorManager) Fetcher {
+func NewEdgeFetcher(ctx context.Context, bucketName string, storeWatcher store.WatermarkStoreWatcher, manager *processor.ProcessorManager, fromBufferPartitionCount int) Fetcher {
 	log := logging.FromContext(ctx).With("bucketName", bucketName)
 	log.Info("Creating a new edge watermark fetcher")
+	var lastProcessedWm []int64
+
+	// to track the last processed watermark for each partition.
+	for i := 0; i < fromBufferPartitionCount; i++ {
+		lastProcessedWm = append(lastProcessedWm, -1)
+	}
+
 	return &edgeFetcher{
 		ctx:              ctx,
 		bucketName:       bucketName,
 		storeWatcher:     storeWatcher,
 		processorManager: manager,
+		lastProcessedWm:  lastProcessedWm,
 		log:              log,
 	}
 }
 
-// GetWatermark gets the smallest watermark for the given offset
-func (e *edgeFetcher) GetWatermark(inputOffset isb.Offset) wmb.Watermark {
+// GetWatermark gets the smallest watermark for the given offset and partition from all the active processors.
+// deletes the processor if it's not active.
+func (e *edgeFetcher) GetWatermark(inputOffset isb.Offset, fromPartitionIdx int32) wmb.Watermark {
+	e.Lock()
+	defer e.Unlock()
```

Comment:
do we need a coarse lock, can't it be granular only where are updating the critical section.